Assets/Tests/EditMode/AuthTests/GoogleCredentialProviderTests.cs

```diff
@@ -0,0 +1,148 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Moq;
+using NUnit.Framework;
+using Src.AuthController.CredentialProviders.Firebase.Google;
+using Src.AuthController.CredentialProviders.Firebase.Google.GoogleRestRequestsAdapter;
+using Src.AuthController.CredentialProviders.Firebase.Google.TokenValidator;
+using Src.AuthController.CredentialProviders.Firebase.Google.UrlOpening;
+using Src.AuthController.REST.PortListener;
+using Src.AuthController.REST.REST_Request_Proxies;
+using Src.AuthController.REST.REST_Response_DTOs;
+namespace Tests.EditMode.AuthTests
+{
+    public class GoogleCredentialProviderTests
+    {
+        [Test]
+        public async Task GetCredentialAsync_ShouldCreateCredentials()
+        {
+            var expectedGoogleIdResponse = new GoogleIdTokenResponse();
+
+            var googleAccessTokenValidatorMock = new Mock<IGoogleAccessTokenValidator>();
+            
+            //Adapter always marks passed TCS as completed and gives it expected response as a result
+            var googleRestRequestsAdapterMock = new Mock<IGoogleRestRequestsAdapter>();
+            googleRestRequestsAdapterMock.Setup(
+                    a => a.ExchangeAuthCodeWithIdToken(
+                        It.IsAny<GoogleIdTokenRequestDtoProxy>(),
+                        It.IsAny<TaskCompletionSource<GoogleIdTokenResponse>>()))
+                .Callback<GoogleIdTokenRequestDtoProxy, TaskCompletionSource<GoogleIdTokenResponse>>(
+                    (dict, tcs) => tcs.SetResult(expectedGoogleIdResponse));
+            
+            var oAuthUrlMock = new Mock<IGoogleOAuthUrl>();
+            
+            //Port listener immediately invokes standard callback
+            var localHttpPortListenerMock = new Mock<ILocalHttpPortListener>();
+            localHttpPortListenerMock.Setup(a => a.StartListening(It.IsAny<Action<string>>()))
+                .Callback<Action<string>>(s => s?.Invoke(""));
+
+            //Test object
+            var googleCredentialProvider = new GoogleCredentialProvider(
+                googleAccessTokenValidatorMock.Object,
+                googleRestRequestsAdapterMock.Object,
+                oAuthUrlMock.Object,
+                localHttpPortListenerMock.Object);
+
+            var result = await googleCredentialProvider.GetCredentialAsync();
+            Assert.AreSame(expectedGoogleIdResponse, result);
+        }
+        
+        [Test]
+        public async Task GetCredentialAsync_ShouldRefreshCredentials()
```

Comment:
Suggested name for this test is GetCredentialAsync_ShouldRefreshCredentials_IfTheyAreNotValid. It is useful to mention condition of class behaviour in the test name.